tests/ModelTest.php

```diff
@@ -1086,21 +1107,21 @@ public function testCreateOrFirst()
         $this->assertTrue($user3->wasRecentlyCreated);
         $this->assertEquals($createdAt, $user1->created_at->getTimestamp());
         $this->assertEquals($createdAt, $user1->updated_at->getTimestamp());
+        $this->assertEquals(['saving', 'creating', 'created', 'saved'], $events);
 
+        $events = [];
         $user4 = User::createOrFirst(
             ['name' => 'Robert Doe'],
             ['name' => 'Maria Doe', 'email' => 'maria.doe@example.com'],
         );
 
         $this->assertSame('Maria Doe', $user4->name);
         $this->assertTrue($user4->wasRecentlyCreated);
-    }
+        $this->assertEquals(['saving', 'creating', 'created', 'saved'], $events);
```

Comment:
Noted that these events are asserted _before_ the transaction commits. I assume that means they're not reliable for inferring that changes were actually persisted to the database.

This reminded me of a PR @alcaeus recently worked on in Doctrine ODM re: lifecycle events and transactions.

Reply:
There is no commit or UoW in Eloquent. Every operation is sent to the DB directly, and events are dispatched in the process, event if the transaction is rollback later.